<!doctype html>
<html lang="en">
<head>
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1" />
<meta name="generator" content="pdoc 0.10.0" />
<title>typing_validation.validation_failure API documentation</title>
<meta name="description" content="Validation failure tracking." />
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/sanitize.min.css" integrity="sha256-PK9q560IAAa6WVRRh76LtCaI8pjTJ2z11v0miyNNjrs=" crossorigin>
<link rel="preload stylesheet" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/10up-sanitize.css/11.0.1/typography.min.css" integrity="sha256-7l/o7C8jubJiy74VsKTidCy1yBkRtiUGbVkYBylBqUg=" crossorigin>
<link rel="stylesheet preload" as="style" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/styles/github.min.css" crossorigin>
<style>:root{--highlight-color:#fe9}.flex{display:flex !important}body{line-height:1.5em}#content{padding:20px}#sidebar{padding:30px;overflow:hidden}#sidebar > *:last-child{margin-bottom:2cm}.http-server-breadcrumbs{font-size:130%;margin:0 0 15px 0}#footer{font-size:.75em;padding:5px 30px;border-top:1px solid #ddd;text-align:right}#footer p{margin:0 0 0 1em;display:inline-block}#footer p:last-child{margin-right:30px}h1,h2,h3,h4,h5{font-weight:300}h1{font-size:2.5em;line-height:1.1em}h2{font-size:1.75em;margin:1em 0 .50em 0}h3{font-size:1.4em;margin:25px 0 10px 0}h4{margin:0;font-size:105%}h1:target,h2:target,h3:target,h4:target,h5:target,h6:target{background:var(--highlight-color);padding:.2em 0}a{color:#058;text-decoration:none;transition:color .3s ease-in-out}a:hover{color:#e82}.title code{font-weight:bold}h2[id^="header-"]{margin-top:2em}.ident{color:#900}pre code{background:#f8f8f8;font-size:.8em;line-height:1.4em}code{background:#f2f2f1;padding:1px 4px;overflow-wrap:break-word}h1 code{background:transparent}pre{background:#f8f8f8;border:0;border-top:1px solid #ccc;border-bottom:1px solid #ccc;margin:1em 0;padding:1ex}#http-server-module-list{display:flex;flex-flow:column}#http-server-module-list div{display:flex}#http-server-module-list dt{min-width:10%}#http-server-module-list p{margin-top:0}.toc ul,#index{list-style-type:none;margin:0;padding:0}#index code{background:transparent}#index h3{border-bottom:1px solid #ddd}#index ul{padding:0}#index h4{margin-top:.6em;font-weight:bold}@media (min-width:200ex){#index .two-column{column-count:2}}@media (min-width:300ex){#index .two-column{column-count:3}}dl{margin-bottom:2em}dl dl:last-child{margin-bottom:4em}dd{margin:0 0 1em 3em}#header-classes + dl > dd{margin-bottom:3em}dd dd{margin-left:2em}dd p{margin:10px 0}.name{background:#eee;font-weight:bold;font-size:.85em;padding:5px 10px;display:inline-block;min-width:40%}.name:hover{background:#e0e0e0}dt:target .name{background:var(--highlight-color)}.name > span:first-child{white-space:nowrap}.name.class > span:nth-child(2){margin-left:.4em}.inherited{color:#999;border-left:5px solid #eee;padding-left:1em}.inheritance em{font-style:normal;font-weight:bold}.desc h2{font-weight:400;font-size:1.25em}.desc h3{font-size:1em}.desc dt code{background:inherit}.source summary,.git-link-div{color:#666;text-align:right;font-weight:400;font-size:.8em;text-transform:uppercase}.source summary > *{white-space:nowrap;cursor:pointer}.git-link{color:inherit;margin-left:1em}.source pre{max-height:500px;overflow:auto;margin:0}.source pre code{font-size:12px;overflow:visible}.hlist{list-style:none}.hlist li{display:inline}.hlist li:after{content:',\2002'}.hlist li:last-child:after{content:none}.hlist .hlist{display:inline;padding-left:1em}img{max-width:100%}td{padding:0 .5em}.admonition{padding:.1em .5em;margin-bottom:1em}.admonition-title{font-weight:bold}.admonition.note,.admonition.info,.admonition.important{background:#aef}.admonition.todo,.admonition.versionadded,.admonition.tip,.admonition.hint{background:#dfd}.admonition.warning,.admonition.versionchanged,.admonition.deprecated{background:#fd4}.admonition.error,.admonition.danger,.admonition.caution{background:lightpink}</style>
<style media="screen and (min-width: 700px)">@media screen and (min-width:700px){#sidebar{width:30%;height:100vh;overflow:auto;position:sticky;top:0}#content{width:70%;max-width:100ch;padding:3em 4em;border-left:1px solid #ddd}pre code{font-size:1em}.item .name{font-size:1em}main{display:flex;flex-direction:row-reverse;justify-content:flex-end}.toc ul ul,#index ul{padding-left:1.5em}.toc > ul > li{margin-top:.5em}}</style>
<style media="print">@media print{#sidebar h1{page-break-before:always}.source{display:none}}@media print{*{background:transparent !important;color:#000 !important;box-shadow:none !important;text-shadow:none !important}a[href]:after{content:" (" attr(href) ")";font-size:90%}a[href][title]:after{content:none}abbr[title]:after{content:" (" attr(title) ")"}.ir a:after,a[href^="javascript:"]:after,a[href^="#"]:after{content:""}pre,blockquote{border:1px solid #999;page-break-inside:avoid}thead{display:table-header-group}tr,img{page-break-inside:avoid}img{max-width:100% !important}@page{margin:0.5cm}p,h2,h3{orphans:3;widows:3}h1,h2,h3,h4,h5,h6{page-break-after:avoid}}</style>
<script async src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.7/latest.js?config=TeX-AMS_CHTML" integrity="sha256-kZafAc6mZvK3W3v1pHOcUix30OHQN6pU/NO2oFkqZVw=" crossorigin></script>
<script defer src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.1.1/highlight.min.js" integrity="sha256-Uv3H6lx7dJmRfRvH8TH6kJD1TSK1aFcwgx+mdg3epi8=" crossorigin></script>
<script>window.addEventListener('DOMContentLoaded', () => hljs.initHighlighting())</script>
</head>
<body>
<main>
<article id="content">
<header>
<h1 class="title">Module <code>typing_validation.validation_failure</code></h1>
</header>
<section id="section-intro">
<p>Validation failure tracking.</p>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">&#34;&#34;&#34;
    Validation failure tracking.
&#34;&#34;&#34;

import sys
import typing
from typing import Any, Optional

def _indent(msg: str) -&gt; str:
    &#34;&#34;&#34; Indent a block of text (possibly with newlines) &#34;&#34;&#34;
    ind = &#34; &#34;*2
    return ind+msg.replace(&#34;\n&#34;, &#34;\n&#34;+ind)

_Acc = typing.TypeVar(&#34;_Acc&#34;)
_T = typing.TypeVar(&#34;_T&#34;, bound=&#34;ValidationFailure&#34;)

class ValidationFailure:
    &#34;&#34;&#34;
        Simple container class for validation failures.
    &#34;&#34;&#34;

    _val: Any
    _t: Any
    _causes: typing.Tuple[&#34;ValidationFailure&#34;, ...]
    _is_union: bool

    def __new__(cls: typing.Type[_T],
                val: Any, t: Any,
                *causes: &#34;ValidationFailure&#34;,
                is_union: bool = False) -&gt; _T:
        instance: _T = super().__new__(cls)
        instance._val = val
        instance._t = t
        instance._causes = causes
        instance._is_union = is_union
        if is_union:
            assert all(cause.val == val for cause in causes)
        return instance

    @property
    def val(self) -&gt; Any:
        &#34;&#34;&#34; The value involved in the validation failure. &#34;&#34;&#34;
        return self._val

    @property
    def t(self) -&gt; Any:
        &#34;&#34;&#34; The type involved in the validation failure. &#34;&#34;&#34;
        return self._t

    @property
    def causes(self) -&gt; typing.Tuple[&#34;ValidationFailure&#34;, ...]:
        &#34;&#34;&#34; Validation failure that in turn caused this failure (if any). &#34;&#34;&#34;
        return self._causes

    @property
    def is_union(self) -&gt; bool:
        &#34;&#34;&#34; Whether this validation failure concerns a union type. &#34;&#34;&#34;
        return self._is_union

    def visit(self, fun: typing.Callable[[Any, Any, _Acc], _Acc], acc: _Acc) -&gt; None:
        &#34;&#34;&#34;
            Performs a pre-order visit of the validation failure tree:

            1. applies `fun(self.val, self.t, acc)` to the failure,
            2. saves the return value as `new_acc`
            3. recurses on all causes using `new_acc`.

            Example usage to pretty-print the validation failure tree using [rich](https://github.com/willmcgugan/rich):

            ```py
            Python 3.9.7
            &gt;&gt;&gt; import rich
            &gt;&gt;&gt; from typing import Union, Collection
            &gt;&gt;&gt; from typing_validation import validate, latest_validation_failure
            &gt;&gt;&gt; validate([[0, 1, 2], {&#34;hi&#34;: 0}], list[Union[Collection[int], dict[str, str]]])
            TypeError: ...
            &gt;&gt;&gt; failure_tree = rich.tree.Tree(&#34;Failure tree&#34;)
            &gt;&gt;&gt; def tree_builder(val, t, tree_tip) -&gt; None:
            ...     label = rich.text.Text(f&#34;({repr(t)}, {repr(val)})&#34;)
            ...     return tree_tip.add(label) # see https://rich.readthedocs.io/en/latest/tree.html
            ...
            &gt;&gt;&gt; latest_validation_failure().visit(tree_builder, failure_tree)
            &gt;&gt;&gt; rich.print(failure_tree)
            Failure tree
            └── (list[typing.Union[typing.Collection[int], dict[str, str]]], [[0, 1, 2], {&#39;hi&#39;: 0}])
                └── (typing.Union[typing.Collection[int], dict[str, str]], {&#39;hi&#39;: 0})
                    ├── (typing.Collection[int], {&#39;hi&#39;: 0})
                    │   └── (&lt;class &#39;int&#39;&gt;, &#39;hi&#39;)
                    └── (dict[str, str], {&#39;hi&#39;: 0})
                        └── (&lt;class &#39;str&#39;&gt;, 0)
            ```
        &#34;&#34;&#34;
        new_acc = fun(self.val, self.t, acc)
        for cause in self.causes:
            cause.visit(fun, new_acc)

    def __str__(self) -&gt; str:
        msg = f&#34;For type {repr(self.t)}, invalid value: {repr(self.val)}&#34;
        if self._is_union:
            for cause in (cause for cause in self.causes if cause.causes):
                msg += &#34;\n&#34;+_indent(f&#34;Detailed failures for member type {repr(cause.t)}:&#34;)
                for sub_cause in cause.causes:
                    msg += &#34;\n&#34;+_indent(_indent(str(sub_cause)))
        else:
            for cause in self.causes:
                msg += &#34;\n&#34;+_indent(str(cause))
        return msg

    def __repr__(self) -&gt; str:
        causes_str = &#34;&#34;
        if self.causes:
            causes_str = &#34;, &#34;+&#34;, &#34;.join(repr(cause) for cause in self.causes)
        is_union_str = &#34;&#34;
        if self._is_union:
            is_union_str = &#34;, is_union=True&#34;
        return f&#34;ValidationFailure({repr(self.val)}, {repr(self.t)}{causes_str}{is_union_str})&#34;

def get_validation_failure(err: TypeError) -&gt; Optional[ValidationFailure]:
    &#34;&#34;&#34;
        Programmatic access to the validation failure tree for the latest validation call.
        Must be called on the type error raised by `validate`.

        ```py
        Python 3.9.7
        &gt;&gt;&gt; from typing_validation import validate, get_validation_failure
        &gt;&gt;&gt; try:
        ...     validate([[0, 1], [1, 2], [2, &#34;hi&#34;]], list[list[int]])
        ... except TypeError as err:
        ...     validation_failure = get_validation_failure(err)
        ...
        &gt;&gt;&gt; validation_failure
        ValidationFailure([[0, 1], [1, 2], [2, &#39;hi&#39;]], list[list[int]],
            ValidationFailure([2, &#39;hi&#39;], list[int],
                ValidationFailure(&#39;hi&#39;, &lt;class &#39;int&#39;&gt;)))
        ```
    &#34;&#34;&#34;
    if not isinstance(err, TypeError):
        raise TypeError(f&#34;Expected TypeError, found {type(err)}&#34;)
    if not hasattr(err, &#34;validation_failure&#34;):
        return None
    validation_failure = getattr(err, &#34;validation_failure&#34;)
    if not isinstance(validation_failure, ValidationFailure):
        return None
    return validation_failure

def latest_validation_failure() -&gt; Optional[ValidationFailure]:
    &#34;&#34;&#34;
        Programmatic access to the validation failure tree for the latest validation call.
        Uses `sys.last_value()`, so it must be called immediately after the error occurred.

        ```py
        Python 3.9.7
        &gt;&gt;&gt; from typing_validation import validate, latest_validation_failure
        &gt;&gt;&gt; validate([[0, 1], [1, 2], [2, &#34;hi&#34;]], list[list[int]])
        TypeError: ...
        &gt;&gt;&gt; latest_validation_failure()
        ValidationFailure([[0, 1], [1, 2], [2, &#39;hi&#39;]], list[list[int]],
            ValidationFailure([2, &#39;hi&#39;], list[int],
                ValidationFailure(&#39;hi&#39;, &lt;class &#39;int&#39;&gt;)))
        ```
    &#34;&#34;&#34;
    try:
        err = sys.last_value # pylint: disable = no-member
    except AttributeError:
        return None
    if not isinstance(err, TypeError):
        return None
    return get_validation_failure(err)</code></pre>
</details>
</section>
<section>
</section>
<section>
</section>
<section>
<h2 class="section-title" id="header-functions">Functions</h2>
<dl>
<dt id="typing_validation.validation_failure.get_validation_failure"><code class="name flex">
<span>def <span class="ident">get_validation_failure</span></span>(<span>err: TypeError) ‑> Optional[<a title="typing_validation.validation_failure.ValidationFailure" href="#typing_validation.validation_failure.ValidationFailure">ValidationFailure</a>]</span>
</code></dt>
<dd>
<div class="desc"><p>Programmatic access to the validation failure tree for the latest validation call.
Must be called on the type error raised by <code>validate</code>.</p>
<pre><code class="language-py">Python 3.9.7
&gt;&gt;&gt; from typing_validation import validate, get_validation_failure
&gt;&gt;&gt; try:
...     validate([[0, 1], [1, 2], [2, &quot;hi&quot;]], list[list[int]])
... except TypeError as err:
...     validation_failure = get_validation_failure(err)
...
&gt;&gt;&gt; validation_failure
ValidationFailure([[0, 1], [1, 2], [2, 'hi']], list[list[int]],
    ValidationFailure([2, 'hi'], list[int],
        ValidationFailure('hi', &lt;class 'int'&gt;)))
</code></pre></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def get_validation_failure(err: TypeError) -&gt; Optional[ValidationFailure]:
    &#34;&#34;&#34;
        Programmatic access to the validation failure tree for the latest validation call.
        Must be called on the type error raised by `validate`.

        ```py
        Python 3.9.7
        &gt;&gt;&gt; from typing_validation import validate, get_validation_failure
        &gt;&gt;&gt; try:
        ...     validate([[0, 1], [1, 2], [2, &#34;hi&#34;]], list[list[int]])
        ... except TypeError as err:
        ...     validation_failure = get_validation_failure(err)
        ...
        &gt;&gt;&gt; validation_failure
        ValidationFailure([[0, 1], [1, 2], [2, &#39;hi&#39;]], list[list[int]],
            ValidationFailure([2, &#39;hi&#39;], list[int],
                ValidationFailure(&#39;hi&#39;, &lt;class &#39;int&#39;&gt;)))
        ```
    &#34;&#34;&#34;
    if not isinstance(err, TypeError):
        raise TypeError(f&#34;Expected TypeError, found {type(err)}&#34;)
    if not hasattr(err, &#34;validation_failure&#34;):
        return None
    validation_failure = getattr(err, &#34;validation_failure&#34;)
    if not isinstance(validation_failure, ValidationFailure):
        return None
    return validation_failure</code></pre>
</details>
</dd>
<dt id="typing_validation.validation_failure.latest_validation_failure"><code class="name flex">
<span>def <span class="ident">latest_validation_failure</span></span>(<span>) ‑> Optional[<a title="typing_validation.validation_failure.ValidationFailure" href="#typing_validation.validation_failure.ValidationFailure">ValidationFailure</a>]</span>
</code></dt>
<dd>
<div class="desc"><p>Programmatic access to the validation failure tree for the latest validation call.
Uses <code>sys.last_value()</code>, so it must be called immediately after the error occurred.</p>
<pre><code class="language-py">Python 3.9.7
&gt;&gt;&gt; from typing_validation import validate, latest_validation_failure
&gt;&gt;&gt; validate([[0, 1], [1, 2], [2, &quot;hi&quot;]], list[list[int]])
TypeError: ...
&gt;&gt;&gt; latest_validation_failure()
ValidationFailure([[0, 1], [1, 2], [2, 'hi']], list[list[int]],
    ValidationFailure([2, 'hi'], list[int],
        ValidationFailure('hi', &lt;class 'int'&gt;)))
</code></pre></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def latest_validation_failure() -&gt; Optional[ValidationFailure]:
    &#34;&#34;&#34;
        Programmatic access to the validation failure tree for the latest validation call.
        Uses `sys.last_value()`, so it must be called immediately after the error occurred.

        ```py
        Python 3.9.7
        &gt;&gt;&gt; from typing_validation import validate, latest_validation_failure
        &gt;&gt;&gt; validate([[0, 1], [1, 2], [2, &#34;hi&#34;]], list[list[int]])
        TypeError: ...
        &gt;&gt;&gt; latest_validation_failure()
        ValidationFailure([[0, 1], [1, 2], [2, &#39;hi&#39;]], list[list[int]],
            ValidationFailure([2, &#39;hi&#39;], list[int],
                ValidationFailure(&#39;hi&#39;, &lt;class &#39;int&#39;&gt;)))
        ```
    &#34;&#34;&#34;
    try:
        err = sys.last_value # pylint: disable = no-member
    except AttributeError:
        return None
    if not isinstance(err, TypeError):
        return None
    return get_validation_failure(err)</code></pre>
</details>
</dd>
</dl>
</section>
<section>
<h2 class="section-title" id="header-classes">Classes</h2>
<dl>
<dt id="typing_validation.validation_failure.ValidationFailure"><code class="flex name class">
<span>class <span class="ident">ValidationFailure</span></span>
<span>(</span><span>val: Any, t: Any, *causes: <a title="typing_validation.validation_failure.ValidationFailure" href="#typing_validation.validation_failure.ValidationFailure">ValidationFailure</a>, is_union: bool = False)</span>
</code></dt>
<dd>
<div class="desc"><p>Simple container class for validation failures.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">class ValidationFailure:
    &#34;&#34;&#34;
        Simple container class for validation failures.
    &#34;&#34;&#34;

    _val: Any
    _t: Any
    _causes: typing.Tuple[&#34;ValidationFailure&#34;, ...]
    _is_union: bool

    def __new__(cls: typing.Type[_T],
                val: Any, t: Any,
                *causes: &#34;ValidationFailure&#34;,
                is_union: bool = False) -&gt; _T:
        instance: _T = super().__new__(cls)
        instance._val = val
        instance._t = t
        instance._causes = causes
        instance._is_union = is_union
        if is_union:
            assert all(cause.val == val for cause in causes)
        return instance

    @property
    def val(self) -&gt; Any:
        &#34;&#34;&#34; The value involved in the validation failure. &#34;&#34;&#34;
        return self._val

    @property
    def t(self) -&gt; Any:
        &#34;&#34;&#34; The type involved in the validation failure. &#34;&#34;&#34;
        return self._t

    @property
    def causes(self) -&gt; typing.Tuple[&#34;ValidationFailure&#34;, ...]:
        &#34;&#34;&#34; Validation failure that in turn caused this failure (if any). &#34;&#34;&#34;
        return self._causes

    @property
    def is_union(self) -&gt; bool:
        &#34;&#34;&#34; Whether this validation failure concerns a union type. &#34;&#34;&#34;
        return self._is_union

    def visit(self, fun: typing.Callable[[Any, Any, _Acc], _Acc], acc: _Acc) -&gt; None:
        &#34;&#34;&#34;
            Performs a pre-order visit of the validation failure tree:

            1. applies `fun(self.val, self.t, acc)` to the failure,
            2. saves the return value as `new_acc`
            3. recurses on all causes using `new_acc`.

            Example usage to pretty-print the validation failure tree using [rich](https://github.com/willmcgugan/rich):

            ```py
            Python 3.9.7
            &gt;&gt;&gt; import rich
            &gt;&gt;&gt; from typing import Union, Collection
            &gt;&gt;&gt; from typing_validation import validate, latest_validation_failure
            &gt;&gt;&gt; validate([[0, 1, 2], {&#34;hi&#34;: 0}], list[Union[Collection[int], dict[str, str]]])
            TypeError: ...
            &gt;&gt;&gt; failure_tree = rich.tree.Tree(&#34;Failure tree&#34;)
            &gt;&gt;&gt; def tree_builder(val, t, tree_tip) -&gt; None:
            ...     label = rich.text.Text(f&#34;({repr(t)}, {repr(val)})&#34;)
            ...     return tree_tip.add(label) # see https://rich.readthedocs.io/en/latest/tree.html
            ...
            &gt;&gt;&gt; latest_validation_failure().visit(tree_builder, failure_tree)
            &gt;&gt;&gt; rich.print(failure_tree)
            Failure tree
            └── (list[typing.Union[typing.Collection[int], dict[str, str]]], [[0, 1, 2], {&#39;hi&#39;: 0}])
                └── (typing.Union[typing.Collection[int], dict[str, str]], {&#39;hi&#39;: 0})
                    ├── (typing.Collection[int], {&#39;hi&#39;: 0})
                    │   └── (&lt;class &#39;int&#39;&gt;, &#39;hi&#39;)
                    └── (dict[str, str], {&#39;hi&#39;: 0})
                        └── (&lt;class &#39;str&#39;&gt;, 0)
            ```
        &#34;&#34;&#34;
        new_acc = fun(self.val, self.t, acc)
        for cause in self.causes:
            cause.visit(fun, new_acc)

    def __str__(self) -&gt; str:
        msg = f&#34;For type {repr(self.t)}, invalid value: {repr(self.val)}&#34;
        if self._is_union:
            for cause in (cause for cause in self.causes if cause.causes):
                msg += &#34;\n&#34;+_indent(f&#34;Detailed failures for member type {repr(cause.t)}:&#34;)
                for sub_cause in cause.causes:
                    msg += &#34;\n&#34;+_indent(_indent(str(sub_cause)))
        else:
            for cause in self.causes:
                msg += &#34;\n&#34;+_indent(str(cause))
        return msg

    def __repr__(self) -&gt; str:
        causes_str = &#34;&#34;
        if self.causes:
            causes_str = &#34;, &#34;+&#34;, &#34;.join(repr(cause) for cause in self.causes)
        is_union_str = &#34;&#34;
        if self._is_union:
            is_union_str = &#34;, is_union=True&#34;
        return f&#34;ValidationFailure({repr(self.val)}, {repr(self.t)}{causes_str}{is_union_str})&#34;</code></pre>
</details>
<h3>Instance variables</h3>
<dl>
<dt id="typing_validation.validation_failure.ValidationFailure.causes"><code class="name">var <span class="ident">causes</span> : Tuple[<a title="typing_validation.validation_failure.ValidationFailure" href="#typing_validation.validation_failure.ValidationFailure">ValidationFailure</a>, ...]</code></dt>
<dd>
<div class="desc"><p>Validation failure that in turn caused this failure (if any).</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">@property
def causes(self) -&gt; typing.Tuple[&#34;ValidationFailure&#34;, ...]:
    &#34;&#34;&#34; Validation failure that in turn caused this failure (if any). &#34;&#34;&#34;
    return self._causes</code></pre>
</details>
</dd>
<dt id="typing_validation.validation_failure.ValidationFailure.is_union"><code class="name">var <span class="ident">is_union</span> : bool</code></dt>
<dd>
<div class="desc"><p>Whether this validation failure concerns a union type.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">@property
def is_union(self) -&gt; bool:
    &#34;&#34;&#34; Whether this validation failure concerns a union type. &#34;&#34;&#34;
    return self._is_union</code></pre>
</details>
</dd>
<dt id="typing_validation.validation_failure.ValidationFailure.t"><code class="name">var <span class="ident">t</span> : Any</code></dt>
<dd>
<div class="desc"><p>The type involved in the validation failure.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">@property
def t(self) -&gt; Any:
    &#34;&#34;&#34; The type involved in the validation failure. &#34;&#34;&#34;
    return self._t</code></pre>
</details>
</dd>
<dt id="typing_validation.validation_failure.ValidationFailure.val"><code class="name">var <span class="ident">val</span> : Any</code></dt>
<dd>
<div class="desc"><p>The value involved in the validation failure.</p></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">@property
def val(self) -&gt; Any:
    &#34;&#34;&#34; The value involved in the validation failure. &#34;&#34;&#34;
    return self._val</code></pre>
</details>
</dd>
</dl>
<h3>Methods</h3>
<dl>
<dt id="typing_validation.validation_failure.ValidationFailure.visit"><code class="name flex">
<span>def <span class="ident">visit</span></span>(<span>self, fun: Callable[[Any, Any, ~_Acc], ~_Acc], acc: ~_Acc) ‑> None</span>
</code></dt>
<dd>
<div class="desc"><p>Performs a pre-order visit of the validation failure tree:</p>
<ol>
<li>applies <code>fun(self.val, self.t, acc)</code> to the failure,</li>
<li>saves the return value as <code>new_acc</code></li>
<li>recurses on all causes using <code>new_acc</code>.</li>
</ol>
<p>Example usage to pretty-print the validation failure tree using <a href="https://github.com/willmcgugan/rich">rich</a>:</p>
<pre><code class="language-py">Python 3.9.7
&gt;&gt;&gt; import rich
&gt;&gt;&gt; from typing import Union, Collection
&gt;&gt;&gt; from typing_validation import validate, latest_validation_failure
&gt;&gt;&gt; validate([[0, 1, 2], {&quot;hi&quot;: 0}], list[Union[Collection[int], dict[str, str]]])
TypeError: ...
&gt;&gt;&gt; failure_tree = rich.tree.Tree(&quot;Failure tree&quot;)
&gt;&gt;&gt; def tree_builder(val, t, tree_tip) -&gt; None:
...     label = rich.text.Text(f&quot;({repr(t)}, {repr(val)})&quot;)
...     return tree_tip.add(label) # see https://rich.readthedocs.io/en/latest/tree.html
...
&gt;&gt;&gt; latest_validation_failure().visit(tree_builder, failure_tree)
&gt;&gt;&gt; rich.print(failure_tree)
Failure tree
└── (list[typing.Union[typing.Collection[int], dict[str, str]]], [[0, 1, 2], {'hi': 0}])
    └── (typing.Union[typing.Collection[int], dict[str, str]], {'hi': 0})
        ├── (typing.Collection[int], {'hi': 0})
        │   └── (&lt;class 'int'&gt;, 'hi')
        └── (dict[str, str], {'hi': 0})
            └── (&lt;class 'str'&gt;, 0)
</code></pre></div>
<details class="source">
<summary>
<span>Expand source code</span>
</summary>
<pre><code class="python">def visit(self, fun: typing.Callable[[Any, Any, _Acc], _Acc], acc: _Acc) -&gt; None:
    &#34;&#34;&#34;
        Performs a pre-order visit of the validation failure tree:

        1. applies `fun(self.val, self.t, acc)` to the failure,
        2. saves the return value as `new_acc`
        3. recurses on all causes using `new_acc`.

        Example usage to pretty-print the validation failure tree using [rich](https://github.com/willmcgugan/rich):

        ```py
        Python 3.9.7
        &gt;&gt;&gt; import rich
        &gt;&gt;&gt; from typing import Union, Collection
        &gt;&gt;&gt; from typing_validation import validate, latest_validation_failure
        &gt;&gt;&gt; validate([[0, 1, 2], {&#34;hi&#34;: 0}], list[Union[Collection[int], dict[str, str]]])
        TypeError: ...
        &gt;&gt;&gt; failure_tree = rich.tree.Tree(&#34;Failure tree&#34;)
        &gt;&gt;&gt; def tree_builder(val, t, tree_tip) -&gt; None:
        ...     label = rich.text.Text(f&#34;({repr(t)}, {repr(val)})&#34;)
        ...     return tree_tip.add(label) # see https://rich.readthedocs.io/en/latest/tree.html
        ...
        &gt;&gt;&gt; latest_validation_failure().visit(tree_builder, failure_tree)
        &gt;&gt;&gt; rich.print(failure_tree)
        Failure tree
        └── (list[typing.Union[typing.Collection[int], dict[str, str]]], [[0, 1, 2], {&#39;hi&#39;: 0}])
            └── (typing.Union[typing.Collection[int], dict[str, str]], {&#39;hi&#39;: 0})
                ├── (typing.Collection[int], {&#39;hi&#39;: 0})
                │   └── (&lt;class &#39;int&#39;&gt;, &#39;hi&#39;)
                └── (dict[str, str], {&#39;hi&#39;: 0})
                    └── (&lt;class &#39;str&#39;&gt;, 0)
        ```
    &#34;&#34;&#34;
    new_acc = fun(self.val, self.t, acc)
    for cause in self.causes:
        cause.visit(fun, new_acc)</code></pre>
</details>
</dd>
</dl>
</dd>
</dl>
</section>
</article>
<nav id="sidebar">
<h1>Index</h1>
<div class="toc">
<ul></ul>
</div>
<ul id="index">
<li><h3>Super-module</h3>
<ul>
<li><code><a title="typing_validation" href="index.html">typing_validation</a></code></li>
</ul>
</li>
<li><h3><a href="#header-functions">Functions</a></h3>
<ul class="">
<li><code><a title="typing_validation.validation_failure.get_validation_failure" href="#typing_validation.validation_failure.get_validation_failure">get_validation_failure</a></code></li>
<li><code><a title="typing_validation.validation_failure.latest_validation_failure" href="#typing_validation.validation_failure.latest_validation_failure">latest_validation_failure</a></code></li>
</ul>
</li>
<li><h3><a href="#header-classes">Classes</a></h3>
<ul>
<li>
<h4><code><a title="typing_validation.validation_failure.ValidationFailure" href="#typing_validation.validation_failure.ValidationFailure">ValidationFailure</a></code></h4>
<ul class="">
<li><code><a title="typing_validation.validation_failure.ValidationFailure.causes" href="#typing_validation.validation_failure.ValidationFailure.causes">causes</a></code></li>
<li><code><a title="typing_validation.validation_failure.ValidationFailure.is_union" href="#typing_validation.validation_failure.ValidationFailure.is_union">is_union</a></code></li>
<li><code><a title="typing_validation.validation_failure.ValidationFailure.t" href="#typing_validation.validation_failure.ValidationFailure.t">t</a></code></li>
<li><code><a title="typing_validation.validation_failure.ValidationFailure.val" href="#typing_validation.validation_failure.ValidationFailure.val">val</a></code></li>
<li><code><a title="typing_validation.validation_failure.ValidationFailure.visit" href="#typing_validation.validation_failure.ValidationFailure.visit">visit</a></code></li>
</ul>
</li>
</ul>
</li>
</ul>
</nav>
</main>
<footer id="footer">
<p>Generated by <a href="https://pdoc3.github.io/pdoc" title="pdoc: Python API documentation generator"><cite>pdoc</cite> 0.10.0</a>.</p>
</footer>
</body>
</html>